### ISSUE TITLE
metaDataTags specification

### ISSUE THREAD

**Issue by jayd1860** [fNIRS/snirf#32]
Hi Qianqian, 

NOTE: Before I respond to your email just wanted to mention, while we say what is an HDF5 Groups in the spec, we do not explicitly say what is a HDF5 Dataset (maybe we should to avoid confusion, but that's a discussion for a later time). We only imply it by using any non-group basic type such as "numeric" or "string" to mean HDF5 Dataset.  So my answer will  assume that "numeric" or "string"  next to "Type:" is basically an HDF5 Dataset. Now on to my response to your email ...

Your statement: "the entire metaDataTags can be just a single dataset of string ...". 
Yes exactly! If we choose to define metaDataTags as a Dataset of type string, then we simply change Type: from the current "group array" to "2-D string array" like this 

      /nirs/metaDataTags
      Type: 2-D string array 

Regarding the alternative way: Sorry, I didn't mean to imply anything complicated like hash maps (we don't care about quick searches through hash tables for instance). My other proposal was simply to use the same type of definition we use for all the other indexed groups of structured variables in the SNIRF spec. Let's take stim as an example: each element of the stim group contains a single name/data condition. In the SNIRF specification looks like this:

     /nirs(i)/stim(j)
     Type: indexed group
 
     /nirs(i)/stim(j)/name
     Type: string

     /nirs(i)/stim(j)/data
     Type: numeric 2-D array

I'm proposing  we define metaDataTags the same way, as an indexed group with 2 Dataset fields: name and value. So each element of metaDataTags would contain ONE tag. This definition adapted to our SNIRF spec would look like this 

      /nirs/metaDataTags(i)
      Type: indexed group

      /nirs/metaDataTags(i)/name
      Type: string

      /nirs/metaDataTags(i)/value
      Type: string

To be clear, it really does not matter to me which of the 2 ways we use, as long as we address my main point: that the current definition of metaDataTags does NOT define ANY dataset - we only define a group, which in the HDF5 world does not contain basic data (not directly anyway). 

Either of the above proposals solves this issue by defining Datasets: the first as a string array, the second as an indexed group of name/value string Dataset fields. 

Lastly, I'm not clear about one aspect of the "SNIRF data format summary" section (which is otherwise a great overview of the SNIRF data structure !). Under metaDataTags, there's a list of tag values like "SubjectID", "MeasurementDate", "MeasurementTime", "LengthUnit", etc as if they were variable names. But, unless I'm misunderstanding something, these are not variable names, they are tag values (required though they may be). It seems to me, their required status belongs only in the description not in the data structure specification where only variable names belong. 

In any case let me know what you think. 

I would be happy to modify the spec to use one of the above definitions if no one has any objections. 

Jay 

Not sure if @huppertt will get this.

**Comment by fangq** [comment]
@jayd1860 and @dboas, the difference between using a 2D (N*2) string array dataset vs a more complex hdf5 construct (such as a group, or group of groups) boils down to whether you prefer to 

1. use hdf5 path `"/nirs/metaDataTags/keyname"` such as `"/nirs/metaDataTags/SubjectID"`  to access each metadata item, or
2. you want to read out the entire header and do your own search (key-value lookup).

if you prefer the first case (i.e. let hdf5 to do the seeking and selective reading), then I think we will have to use a group for `metaDataTags` and each key as a dataset (or subgroup).

if you think the 2nd way is acceptable, then, I think we can use a 2-D string array, packaged in a single dataset.

I am ok with either way. @huppertt, correct me if my understanding to group/dataset is wrong.

**Comment by jayd1860** [comment]
@fangq, @huppertt, @dboas,

Ok I see now what you mean about searching for keys: I guess you're showing me a 3rd way to do this (method 1. in your post), that is, to use the tag names themselves to name the SNIRF variables. And that seems like a legitimate way to do it as well. The only downside with that method is that you cannot specify a fixed data structure for SNIRF because the variable names themselves are ... variable.  I think that might get confusing because it blurs the distinction between variable names and their data. This is not the case with method 2 (in your post) or with my other proposal where metaDataTags is a Group containing the Datasets name and value. In both of those cases tag names and values are all treated as just data. 

Bottom line is, I am in favor of method 2 (in your post) where metaDataTags is a Dataset of type 2-D string array.

**Comment by dboas** [comment]
I am also in favor of “metaDataTags is a Dataset of type 2-D string array”

**Comment by huppertt** [comment]
The reason I had initially put it as a key/value structure was that I thought values could be any type (e.g. string or numeric or other).  Whereas the 2D string array would require the reader to parse (e.g.) age into a numeric value from the stored string type.  Its way more work for the reader to have to code to check the type of the value in the file and import it that way, but removes ambiguity of what the intended type of the value was.  

I was specifically thinking of my analysis software and how it uses this data in group level statistical models; where strings are treated as categorical covariates in the ANOVA models, but numerics are continuous.   We run into this problem a bunch importing demographic info from excel because excel treats "NaN" as a string and ends up converting the entire column of otherwise numeric values to strings if there is a single NaN entry.  

Bottom line- I am ok with the 2D string, but I did have a preference to the key/value approach.

**Comment by dboas** [comment]
Ted, how do we set up in HDF5 where the value could be string or numeric? I am fine with the 2D string array. It is easy for the programmer to do the conversion as needed.
David

**Comment by huppertt** [comment]
When you create the dataspace location in HDF5, you define the data type (string, int, etc).  The data file saves the type info and when you read it back, you first need to retrieve the data type and alter the read command accordingly.  

The code (from my C# toolbox) is something like:


 private static hid_t AddDataValue(hid_t parentLoc, string name, double data)
        {
            // Write the data to the /../name field
            hid_t type = H5T.NATIVE_DOUBLE;   // change this to alter the data storage type
            hsize_t[] dims = new hsize_t[1];
            dims[0] = 1;

            hid_t spaceId = H5S.create_simple(1, dims, null);
            hid_t dataSetId = H5D.create(parentLoc, name, type, spaceId);

            // Write the integer data to the data set.
            GCHandle hnd = GCHandle.Alloc(data, GCHandleType.Pinned);
            H5D.write(dataSetId, type, H5S.ALL, H5S.ALL, H5P.DEFAULT,hnd.AddrOfPinnedObject());
            hnd.Free();
            H5D.close(dataSetId);
            return dataSetId;
        }

**Comment by fangq** [comment]
@dboas, from what I read (within just a limited time), there are several (probably too many) ways to store heterogeneous data in a hash/map in hdf5.

The cleanest way is to write your own hdf5 constructs.

In the below example, I use the low-level MATLAB hdf5 interface to write a metadata header (`meta` as a struct), containing a string and a numerical subfield, into a sample file.

```
typemap.char='H5T_C_S1';
typemap.double='H5T_IEEE_F64LE';
typemap.single='H5T_IEEE_F32LE';

meta.SubjectID='testsubj';
meta.SubjectAge=20;

pd = 'H5P_DEFAULT';

fid = H5F.create('testmeta7.h5');
gid = H5G.create(fid,'nirs',pd,pd,pd);
mid = H5G.create(gid,'metaDataTags',pd,pd,pd);
metakeys=fieldnames(meta);
for i=1:length(metakeys)
   val=meta.(metakeys{i});
   mv1= H5D.create(mid,metakeys{i},H5T.copy(typemap.(class(val))),H5S.create_simple(ndims(val), fliplr(size(val)),fliplr(size(val))),pd);
   H5D.write(mv1,'H5ML_DEFAULT','H5S_ALL','H5S_ALL',pd,val);
end
H5F.close(fid);
```
if your data contains more types, you can just extend the `typemap` and add `HDT_*` support.

once you write this file, you can then display the structure using
```
>> h5disp('testmeta7.h5')
HDF5 testmeta7.h5 
Group '/' 
    Group '/nirs' 
        Group '/nirs/metaDataTags' 
            Dataset 'SubjectAge' 
                Size:  1x1
                MaxSize:  1x1
                Datatype:   H5T_IEEE_F64LE (double)
                ChunkSize:  []
                Filters:  none
                FillValue:  0.000000
            Dataset 'SubjectID' 
                Size:  1x8
                MaxSize:  1x8
                Datatype:   H5T_STRING
                    String Length: 1
                    Padding: H5T_STR_NULLTERM
                    Character Set: H5T_CSET_ASCII
                    Character Type: H5T_C_S1
                ChunkSize:  []
                Filters:  none
                FillValue:  ' '
```

You can see that the `metaDataTags` is a group, with each item a dataset. The benefit of this construct is that you can define different data types, and then query each data field using their full path.

```
fid = H5F.open('testmeta7.h5');
metaid = H5D.open(fid,'/nirs/metaDataTags/SubjectAge');
type_id = H5D.get_type(metaid)
age = H5D.read(metaid)
H5F.close(fid);
```

Of course, if you do not want to use low-level APIs, for convenience, MATLAB's hdf5 high-level api can also store similar data using a variety of hdf5 containers.

I am posting 3 sample structures that I mentioned in the offline email exchanges. You can see from these examples, the metaDataTags can be 1) a dataset of 2-D string array, 2) a database of a compound data. I am sure there must be other constructs to store similar info.

The question routes back to - in what way you want to query this metadata header, and whether it is necessary to allow heterogeneous data types for each subfield. These are all doable within hdf5.

--------------------

Using 2-D string array in MATLAB
```
a={'name1','val1';'name2','val2'; 'name3','val3'}
hdf5write('testmeta.h5', '/metaDataTags', a');
```
```
fangq@taote:~$ h5dump testmeta.h5
HDF5 "testmeta.h5" {
GROUP "/" {
   DATASET "metaDataTags" {
      DATATYPE  H5T_STRING {
         STRSIZE 6;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_ASCII;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SIMPLE { ( 3, 2 ) / ( 3, 2 ) }
      DATA {
      (0,0): "name1", "val1",
      (1,0): "name2", "val2",
      (2,0): "name3", "val3"
      }
   }
}
```

Saving as a hash/map (key-value pairs):

```
a=struct('name1','val1','name2','val2','name3','val3')
hdf5write('testmeta2.h5', '/metaDataTags', a);
```
```
fangq@taote:~$ h5dump testmeta2.h5
HDF5 "testmeta2.h5" {
GROUP "/" {
   DATASET "metaDataTags" {
      DATATYPE  H5T_COMPOUND {
         H5T_STRING {
            STRSIZE 5;
            STRPAD H5T_STR_NULLTERM;
            CSET H5T_CSET_ASCII;
            CTYPE H5T_C_S1;
         } "name1";
         H5T_STRING {
            STRSIZE 5;
            STRPAD H5T_STR_NULLTERM;
            CSET H5T_CSET_ASCII;
            CTYPE H5T_C_S1;
         } "name2";
         H5T_STRING {
            STRSIZE 5;
            STRPAD H5T_STR_NULLTERM;
            CSET H5T_CSET_ASCII;
            CTYPE H5T_C_S1;
         } "name3";
      }
      DATASPACE  SIMPLE { ( 1 ) / ( 1 ) }
      DATA {
      (0): {
            "val1",
            "val2",
            "val3"
         }
      }
   }
}
}
```

In the above example, I can also define different data types for each key:
```
a=struct('name1','a string','name2',uint8(20),'name3',[2,1])
hdf5write('testmeta3.h5', '/metaDataTags', a);
```
```
fangq@taote:~$ h5dump testmeta3.h5
HDF5 "testmeta3.h5" {
GROUP "/" {
   DATASET "metaDataTags" {
      DATATYPE  H5T_COMPOUND {
         H5T_STRING {
            STRSIZE 9;
            STRPAD H5T_STR_NULLTERM;
            CSET H5T_CSET_ASCII;
            CTYPE H5T_C_S1;
         } "name1";
         H5T_STD_U8LE "name2";
         H5T_ARRAY { [2] H5T_IEEE_F64LE } "name3";
      }
      DATASPACE  SIMPLE { ( 1 ) / ( 1 ) }
      DATA {
      (0): {
            "a string",
            20,
            [ 2, 1 ]
         }
      }
   }
}
}
```

**Comment by fangq** [comment]
To summarize the options we have, we can define

1. group of groups (probably overkill)
```
/nirs/metaDataTags      <- group of groups
          |->SubjectID       <- group
          |      |-> value      <- dataset to store actual per-item data
          \...
```
2. group of datasets
```
/nirs/metaDataTags      <- group
          |->SubjectID       <- dataset
          \...
```
3. a single dataset of compound types
```
/nirs/metaDataTags      <- dataset
          TYPE{"SubjectID", "Age", ... }  <- compound datatype
          DATA{"val1",val2, ... }         <- compound data values
```
4. a single dataset of 2-D string arrays (can not query value per key, limited to string key only)
```
/nirs/metaDataTags      <- dataset
          [
              [ "SubjectID", "subj1"],
              [ "Age", "20"],
              ...
          ]
```

**Comment by dboas** [comment]
I understand that the original intention was

  1.  a single dataset of compound types
/nirs/metaDataTags      <- dataset
          TYPE{"SubjectID", "Age", ... }  <- compound datatype
          DATA{"val1",val2, ... }         <- compound data values

And given that we have code examples for how to implement this, it seems like the way to go.

Jay, are you okay with this?

**Comment by fangq** [comment]
@jayd1860 and @dboas, I made a patch to Jay's Homer3 code a moment ago

https://github.com/BUNPC/Homer3/pull/2

This patch addressed a few issues

1. the `MetaDataTagsClass` now handles all `metaDataTags` fields as a struct, instead of a single key/value pair in Jay's original version. This may significantly lower the overhead of reading/writing files

2. I used the commands in my previous post https://github.com/fNIRS/snirf/issues/32#issuecomment-515521619 to save/load metaDataTags, this allows each subfield to have different types (some can be strings, some can be numbers), provides more flexibility.

3. the output file structure is format-2 in this https://github.com/fNIRS/snirf/issues/32#issuecomment-515526133. David previously preferred format-3 (https://github.com/fNIRS/snirf/issues/32#issuecomment-515703408), I tried it first, but it turned out a compound record pad spaces to all string-valued sub-fields to make them equal in length, this can cause inefficiency in storage. I think format-2 has the same flexibility. I removed `key/value` subfields because I don't think using those extra subfields is necessary. It also makes it easier to describe the format in the specification.

@jayd1860, I tested the updated class using the `demo_snirf.m` script in the Examples folder, and passed all test cases. If you like this, I will revert your changes related to the addition of key/value pairs. Please also update the sample data files in `snirf_samples` repo. 

Happy to discuss if you have questions.

**Comment by fangq** [comment]
here is the dump of the output data file - `/nirs/metaDataTags` is a group, and each subfield, say `/nirs/metaDataTags/SubjectID` is a dataset. Again, this is format-2 in https://github.com/fNIRS/snirf/issues/32#issuecomment-515526133

> fangq@taote$ h5dump -g  /nirs/metaDataTags neuro_run01.snirf
```
HDF5 "neuro_run01.snirf" {
GROUP "/nirs/metaDataTags" {
   DATASET "LengthUnit" {
      DATATYPE  H5T_STRING {
         STRSIZE 1;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_ASCII;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SIMPLE { ( 2, 1 ) / ( 2, 1 ) }
      DATA {
      (0,0): "m",
      (1,0): "m"
      }
   }
   DATASET "MeasurementDate" {
      DATATYPE  H5T_STRING {
         STRSIZE 1;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_ASCII;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SIMPLE { ( 7, 1 ) / ( 7, 1 ) }
      DATA {
      (0,0): "u",
      (1,0): "n",
      (2,0): "k",
      (3,0): "n",
      (4,0): "o",
      (5,0): "w",
      (6,0): "n"
      }
   }
   DATASET "MeasurementTime" {
      DATATYPE  H5T_STRING {
         STRSIZE 1;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_ASCII;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SIMPLE { ( 7, 1 ) / ( 7, 1 ) }
      DATA {
      (0,0): "u",
      (1,0): "n",
      (2,0): "k",
      (3,0): "n",
      (4,0): "o",
      (5,0): "w",
      (6,0): "n"
      }
   }
   DATASET "SubjectID" {
      DATATYPE  H5T_STRING {
         STRSIZE 1;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_ASCII;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SIMPLE { ( 7, 1 ) / ( 7, 1 ) }
      DATA {
      (0,0): "d",
      (1,0): "e",
      (2,0): "f",
      (3,0): "a",
      (4,0): "u",
      (5,0): "l",
      (6,0): "t"
      }
   }
}
}


```

**Comment by jayd1860** [comment]
Hi Qianqian, 

Thanks for the very helpful comments and looking over the Homer3 code. I made the fixes that you caught (thanks!) and made some other changes in response to your great suggestions. I do agree I need to show how to load arrays of objects in Snirf. For this purpose I added some methods to the SnirfClass and a simple demo (demo_snirf_readfile) which shows exactly how to do this and updated the README ( basically you 1. run cd to the Homer3 folder, 2. run setpaths, 3. run demo_snirf.m to generate snirf files in the Examples folder, then run demo_snirf_readfile.m to see exactly how to read all the metaDataTags, and stim fields ). 

I will look at your pull request (I might not get to it until next week) But from a quick glance it seems that the implementation you describe treats metaDataTags differently than other groups like aux, stim, and data. But their specification in the Spec as it currently stands is essentially the same (for example aux and metaDataTags are virtually identical in their structure). Therefore I would prefer not to treat metaDatatags any differently in the Homer3 code with a totally different implementation than for the other fields. 

Basically in the Homer3 Snirf implementation each element type gets it's own class, but to read an array of objects of that type, you have to go through a SnirfClass method for reading/writing arrays (or groups in HDF5 lingo). So for instance AuxClass defines one element of an array of AuxClass objects (AuxClass.load()) and knows how to load one element of that type but only SnirfClass knows how to read an array of these (by invoking the  SnirfClass.LoadAux() method. Same idea for metaDataTags, stim, data, etc. This design is used throughout the Homer3 SnirfClass implementation. 

If you run the demo_snirf_readfile (after running demo_snirf), it shows exactly how to read groups of metaDataTags and stim and illustrates the overall design. 

Hope that makes sense.

**Comment by fangq** [comment]
@jayd1860, I've already run the demos script and compared your original outputs and mine.

here is the output data from your original class:

```
>> ss=h5load('neuro_run01.snirf');
>> ss

ss = 

  struct with fields:

    formatVersion: [5×1 char]
             nirs: [1×1 struct]

>> ss.nirs

ans = 

  struct with fields:

             aux1: [1×1 struct]
            data1: [1×1 struct]
    metaDataTags1: [1×1 struct]
    metaDataTags2: [1×1 struct]
    metaDataTags3: [1×1 struct]
    metaDataTags4: [1×1 struct]
            probe: [1×1 struct]
            stim1: [1×1 struct]
            stim2: [1×1 struct]

>> ss.nirs.metaDataTags1

ans = 

  struct with fields:

     name: [10×1 char]
    value: [5×1 char]
```
here is the output data from my updated MetaDataTagsClass

```
>> ss=h5load('neuro_run01.snirf');
>> ss.nirs

ans = 

  struct with fields:

            aux1: [1×1 struct]
           data1: [1×1 struct]
    metaDataTags: [1×1 struct]
           probe: [1×1 struct]
           stim1: [1×1 struct]
           stim2: [1×1 struct]

>> ss.nirs.metaDataTags

ans = 

  struct with fields:

         LengthUnit: 'mm'
    MeasurementDate: 'unknown'
    MeasurementTime: 'unknown'
          SubjectID: 'default'
```

I think my output is a lot cleaner (and more flexible). It is also a lot easier to look-up values in both MATLAB and in HDF5 file - in your output, one has to loop over all subfields to match the "name" tag (because you don't know the number in the group name), while in my case, I can use /nirs/metaDataTags/keyname to read a value directly.

Also, I expect the metaDataTags subfields will grow significantly once people start using it. The index group approach is very difficult to scale, can cause high overhead in file sizes and read/write speed. I also don't think the key/value is necessary - the key info can be stored in the dataset names, and the value can be the dataset value.

Right now, your class takes about 6 seconds to read/write the FingerTapping sample data on my computer, I believe the frequent file open/close (per data item) contributes to a big part of this overhead. I will show you, in a different post, that this can be cut down significantly (to ~1s) if you only open the file once.

**Comment by fangq** [comment]
Jay, also following your comment, I want to mention that the data stored in `metaDataTags` are indeed different from `aux/stim/data`.

Basically, `aux/stim` are cell arrays (i.e., indexed arrays), but metaDataTags is a structure/map.

it is easier to show the difference in JSON's notation (`'[]'` vs `'{}'`): for `aux`, for example, what you have is

```
"aux":[
    {"name": "..", "time": [],... },
    {"name": "..", "time": [],... },
    ...
]
```

where the elements are accessed by index (1,2,3...), while the metaDataTag elements are accessed by names instead of indices

```
"metaDataTags": {
    "SubjectID": "...",
    "MeasurementDate": "..",
    ...
}
```

**Comment by jayd1860** [comment]
Ah ok Qianqian,

I will look into that. Basically if that is true, then the Homer3 Snirf writer is doing something wrong. It should definitely NOT be storing it differently. 

Let me look into it and get back to you. 

Jay

**Comment by jayd1860** [comment]
Hi Qianqian, 

About your previous lengthy comment, perhaps it would be a lot easier for us to figure this out in person. Do you think this is possible? Maybe I can visit NE next week?

We're going back and forth for a long time now - and it's very time consuming, and it's hard to really appreciate what the other person is saying. 

Let me know if you'd be open to that. 

Jay

**Comment by dboas** [comment]
Qianqian,
That would be great if Jay could come visit you to nail this down.
David

**Comment by fangq** [comment]
close after this pull request https://github.com/fNIRS/snirf/pull/36
